proposals/0003-spec.md

```diff
@@ -0,0 +1,249 @@
+
+
+## Protocol Adaptor
+
+
+Each provider MUST provide the REST/HTTP interaction, but MAY provide other protocols such as TCP or JMX. When possible, the output MUST be the JSON output returned by the equivalent HTTP calls (Appendix B). The request is protocol specific.
+Healthcheck Response information
+
+
+* The primary information MUST be boolean, it needs to be consumed by other machines. Anything between available/unavailable doesn’t make sense or would increase the complexity on the side of the consumer processing that information.
+* The response information MAY contain an additional information holder
+* Consumers MAY process the additional information holder or simply decide to ignore it
+* The response information MUST contain the boolean check state
+* The response information MAY contain the check id (or name)
+Wireformats
+Which format is used to depict the state (Json format as I've seen)
+
+
+* Producer MUST  support JSON encoded payload with simple UP/DOWN states
+* Producers MAY  support an additional information holder with key/value pairs to provide further context (i.e. disk.free.space=120mb).
+* The JSON response payload MUST be compatible with the one described in Appendix B
+* The JSON response MUST contain the `id` entry specifying the name of the check, to support protocols that support external identifier (i.e. URI)
+* The JSON response MUST contain the `result`entry specifying the state as String: “UP” or “DOWN”
+* The JSON MAY support an additional information holder to carry key value pairs that provide additional context
+
+
+# Health Check Procedures
+* A producer MUST support custom health check procedures
+* A producer SHOULD support reasonable out-of-the-box procedures
+* A producer without health check procedures installed MUST returns positive overall outcome (i.e. HTTP 204, no content)
+
+
+## Policies to determine the overall outcome
+When multiple procedures are installed all procedures MUST be executed and the overall outcome needs to be determined.
+* Consumers MUST support a logical conjunction policy to determine the outcome
+* Consumers MUST use the logical conjunction policy by default to determine the outcome
+* Consumers MAY support custom policies to determine the outcome
+Security
+Aspects regarding the secure access of health check information.
+
+
+* A producer MUST enforce security on all check invocations
+* A producer MAY ignore security for trusted origins (i.e. localhost)
+* HTTP Digest Auth MUST be one supported authentication mechanism
+* HTTP Digest Auth SHOULD be the default algorithm for the HTTP protocol binding
+
+# Appendix A: REST interface specifications
+
+| Context       | Verb          | Status Code  | Response |
+| ------------- |:-------------:| :-----:| -----|
+| /health       | GET           | 200, 204, 500, 503 | See Appendix B|
+
+## Status Codes:
+
+* 200 for a health check with a positive outcome
+* 204 in case no health check procedures are installed into the runtime
```

Comment:
the implications here is that there's some discovery mechanism for health checks.  Maybe I missed that, but wouldn't that be required?